system/src/asset_manager_api.cpp

```diff
@@ -0,0 +1,172 @@
+ * License along with this library; if not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include "asset_manager_api.h"
+#include "asset_manager.h"
+#include "check.h"
+#include "scope_guard.h"
+
+using namespace particle;
+
+namespace {
+
+struct Info {
+    spark::Vector<Asset> required;
+    spark::Vector<Asset> available;
+};
+
+void assetToApiAsset(const Asset& asset, asset_manager_asset* apiAsset) {
+    apiAsset->name = asset.name().c_str();
+    apiAsset->size = asset.size();
+    apiAsset->hash_type = asset.hash().type();
+    apiAsset->hash_length = asset.hash().hash().size();
+    apiAsset->hash = asset.hash().hash().data();
+    apiAsset->storage_size = asset.storageSize();
+}
+
+Asset assetFromApiAsset(const asset_manager_asset* apiAsset) {
+    return Asset(apiAsset->name, AssetHash(apiAsset->hash, apiAsset->hash_length, (particle::AssetHash::Type)apiAsset->hash_type));
+}
+
+} // anonymous
+
+int asset_manager_set_notify_hook(asset_manager_notify_hook hook, void* context, void* reserved) {
+    return AssetManager::instance().setNotifyHook(hook, context);
+}
+
+int asset_manager_get_info(asset_manager_info* info, void* reserved) {
+    CHECK_TRUE(info, SYSTEM_ERROR_INVALID_ARGUMENT);
+    auto internal = new Info();
+    CHECK_TRUE(internal, SYSTEM_ERROR_NO_MEMORY);
+    internal->required = AssetManager::instance().requiredAssets();
+    internal->available = AssetManager::instance().availableAssets();
+
+    info->internal = (void*)internal;
+
+    bool ok = false;
+    SCOPE_GUARD({
+        if (!ok) {
+            asset_manager_free_info(info, nullptr);
+        }
+    });
+
+    info->asset_size = sizeof(asset_manager_asset);
+    info->required_count = internal->required.size();
+    info->available_count = internal->available.size();
+
+    if (info->required_count) {
+        info->required = new asset_manager_asset[info->required_count];
+        CHECK_TRUE(info->required, SYSTEM_ERROR_NO_MEMORY);
+        memset(info->required, 0, sizeof(asset_manager_asset) * info->required_count);
+        for (size_t i = 0; i < info->required_count; i++) {
+            assetToApiAsset(internal->required[i], &info->required[i]);
+        }
+    }
+
+    if (info->available_count) {
+        info->available = new asset_manager_asset[info->available_count];
+        CHECK_TRUE(info->available, SYSTEM_ERROR_NO_MEMORY);
+        memset(info->available, 0, sizeof(asset_manager_asset) * info->available_count);
+        for (size_t i = 0; i < info->available_count; i++) {
+            assetToApiAsset(internal->available[i], &info->available[i]);
+        }
+    }
+    ok = true;
+    return 0;
+}
+
+int asset_manager_free_info(asset_manager_info* info, void* reserved) {
+    CHECK_TRUE(info, SYSTEM_ERROR_INVALID_ARGUMENT);
+    if (info->internal) {
+        auto internal = static_cast<Info*>(info->internal);
+        delete internal;
+    }
+    if (info->required) {
+        free(info->required);
```

Comment:
`free()` is used to free the memory allocated with `new`.

Same [here](https://github.com/particle-iot/device-os/pull/2668/files#diff-4c430d849bb9061cdddf693086b5b007baaafbf62d9dd4ef3b87aefcb7619d03R102).